### DISCUSSION TITLE
Yellow/CM4 NVMe support (PCIe/without USB adapter)

### DISCUSSION THREAD

**Discussion by agners** [home-assistant/operating-system#2235] [General]
This thread is meant to collect known not-working NVMes for Yellow or other Raspberry Pi CM4 based boards. It is meant for a more long-term location to collect information previously tracked in https://github.com/home-assistant/operating-system/issues/2178.

A more user friendly version of this information can be found [on the Yellow Support page](https://yellow.home-assistant.io/support/#which-ssds-are-supported).

As different parts of the system talk to the NVMe, communication can fail on different levels. Not all of them have been observed so far, but for completeness this is a (probably incomplete) list of possible failures:

- Type 1: Low-level/Hardware failure (no parts of the system can communicate to the NVMe)
- Type 2: CM4 EEPROM (check with `BOOT_UART=1`) is not able to communicate with the NVMe
- Type 3: CM4 Firmware (start4.elf)  is not able to communicate with the NVMe
- Type 4: HAOS U-Boot is not able to communicate with the NVMe
- Type 5: HAOS Linux is not able to communicate with the NVMe

| Producer                 | Type               | Model             | Failure Type                       | Notes                                                                            |
|---------------------------|---------------------|--------|-------------------------------------|-----------------------------------------------------------------------|
| SK hynix                 | BC711              |          | Type 3                                 | Resolved since RPi firmware 1.20221028 (see https://github.com/raspberrypi/rpi-eeprom/issues/451) |
| WD_BLACK             | SN850X (1TB) | ‎WDS100T2X0E | Type 1               | Low-level enumeration issue (see https://github.com/raspberrypi/linux/issues/5217#issuecomment-1288729514) |
| WD_BLACK              | SN770           |           | Type 1 |                              |
| WD_BLACK              | SN770 SE        |           | Type 1 |                              |
| Crucial                      | P2  | CT250P2SSD8 | Type 3? 4? |                      |             |

Note: Type 2/3/4 failure can still work with a CM4 with eMMC (via data disk feature), as in this setup HAOS boots from eMMC and not from NVMe.

**Comment by agners** [comment]
WD SN770 potentially working (see https://discord.com/channels/330944238910963714/887751499785072711/1066618808795672597).

**Comment by MomentumV** [comment]
I purchased a CT500P2SSD8 and am not having my HA yellow boot. pi CM4 lite.
more console details later.

Edit: CT500P2SSD8 is booting from nvme with a cm4 lite with updated eeprom boot commands

**Reply by MomentumV** [reply]
This seems to have been an out-of-date eeprom problem that did not enumerate nvme as a boot option. I was able to (eventually) get the eeprom updated and it is working now.

**Reply by MomentumV** [reply]
It's particularly troublesome to update the cm4 lite eeprom.

**Comment by agners** [comment]
We've received the following report through support:

> We tried to get the NVMe working on two Yellow v1.2 and with three different Lexar NM620’s, they are visible, but the controller is heavily unstable. Every single installation ends up with a Kernel panic after first boot (found out via Serial), or the system with a Lexar won’t boot at all (leds are stating that there is a “Fatal Firmware Error”).
It’s completely random with the Lexar combination, but all iterations end up with a Kernel panic or failing led status.

**Reply by ivanfrolovau** [reply]
I have a similar issue with Lexar NM620 bought from seeedstudio. The board either boots successfully or shows a "Fatal Firmware Error" (via LEDs) after being powered on. I can't bisect what causes one or the other. After multiple power cycles it finally boots, but the story repeats if you have to reboot.

**Comment by jwillikers** [comment]
Following up here for PR home-assistant/yellow.home-assistant.io#125. I attempted to get a Samsung 980 Pro, model MZ-V8P500, size 500 GB working with the Home Assistant Yellow PoE (v1.3) and was unable to get the installer working. I used a CM4 Lite 8 GB model using the latest `usbboot` firmware. Given the power consumption of the Samsung 980 Pro, I conducted these tests using a 3 amp barrel jack connector instead of PoE power to be sure it wasn't related to the NVMe not getting enough power. I believe I was able to previously boot Raspberry Pi OS off of this NVMe disk on a CM4 IO board using Waveshare's PCIe X1 to M.2 adapter some time ago.

Flashing the latest release of the Home Assistant OS image for the Yellow directly to the NVMe also failed to boot. I was able to verify this same image booted from a USB flash drive by observing the output on the USB serial console. There was no console output when I tried to boot the image off of the NVMe disk.

I did manage to get the installer to install successfully by switching the 980 Pro out for a dram-less 512 GB M.2 2242 Sabrent NVMe I had laying around.
